File: typescript/examples/tools/agents/factory-pattern-agent.ts
#!/usr/bin/env node
/**
 * Matimo Factory Pattern - True AI Agent with Tool Decision Making
 *
 * The agent receives a prompt, uses OpenAI to decide which tool to use,
 * then executes that tool via Matimo.
 *
 * Run: npm run agent:factory
 */

import 'dotenv/config';
import path from 'path';
import { fileURLToPath } from 'url';
import { ChatOpenAI } from '@langchain/openai';
import { MatimoInstance } from 'matimo';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

/**
 * Factory Pattern Agent - Uses AI to decide which tool to call
 */
class FactoryPatternAgent {
  private matimo: MatimoInstance;
  private llm: ChatOpenAI;

  constructor(matimo: MatimoInstance, llm: ChatOpenAI) {
    this.matimo = matimo;
    this.llm = llm;
  }

  /**
   * Generate OpenAI function schemas from Matimo tool definitions
   * Single source of truth - schemas come from tool YAML, not duplicated here
   */
  private getToolSchemas() {
    return this.matimo.listTools().map((tool) => ({
      type: 'function',
      function: {
        name: tool.name,
        description: tool.description,
        parameters: {
          type: 'object',
          properties: Object.entries(tool.parameters || {}).reduce(
            (acc, [paramName, param]) => ({
              ...acc,
              [paramName]: {
                type: param.type,
                enum: param.enum,
                description: param.description,
              },
            }),
            {}
          ),
          required: Object.entries(tool.parameters || {})
            .filter(([_, param]) => param.required)
            .map(([paramName]) => paramName),
        },
      },
    }));
  }

  /**
   * Process a prompt - AI decides which tool to use
   */
  async process(prompt: string): Promise<void> {
    console.log(`\n❓ Prompt: "${prompt}"`);

    try {
      // Prepare system message for tool calling
      const systemMessage = `You are an AI assistant with access to tools. 
Based on the user's request, decide which tool to use and extract the required parameters.
Extract the tool name and parameters, then respond with JSON.`;

      // Get tool schemas dynamically from Matimo
      const toolSchemas = this.getToolSchemas();

      // Create a message with proper formatting for function calling
      const messages = [
        {
          type: 'system',
          content: systemMessage,
        },
        {
          type: 'human',
          content: `User request: ${prompt}\n\nAvailable tools: ${toolSchemas.map((t) => `${t.function.name}: ${t.function.description}`).join(', ')}\n\nRespond with JSON: {"tool": "<tool_name>", "parameters": {...}}`,
        },
      ];

      // Call LLM with tool definitions
      const response = await this.llm.invoke(messages as any);

      // Parse the LLM response
      let toolName: string | null = null;
      let toolParams: Record<string, unknown> | null = null;

      // Try to extract tool call from response
      const content = response.content;

      if (typeof content === 'string') {
        // Try to parse as JSON
        try {
          const parsed = JSON.parse(content);
          toolName = parsed.tool || parsed.function?.name;
          toolParams = parsed.parameters || parsed.function?.parameters || parsed;
        } catch {
          // If not JSON, try to extract from text
          const jsonMatch = content.match(/\{[^{}]*"tool"[^{}]*\}/);
          if (jsonMatch) {
            try {
              const parsed = JSON.parse(jsonMatch[0]);
              toolName = parsed.tool;
              toolParams = parsed.parameters;
            } catch {
              // Continue without params
            }
          }
        }
      }

      // If we found a tool, execute it
      if (toolName && toolParams) {
        await this.executeTool(toolName, toolParams);
      } else {
        console.log(`\n⚠️  No tool call detected in response`);
        console.log(
          `Response: ${typeof content === 'string' ? content.substring(0, 200) : content}`
        );
      }
    } catch (error) {
      console.error(`\n❌ Error: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Execute a tool via Matimo
   * Normalize parameters to match tool schema expectations
   */
  private async executeTool(toolName: string, params: Record<string, unknown>): Promise<void> {
    try {
      const tool = this.matimo.getTool(toolName);
      if (!tool) {
        console.log(`\n❌ Tool '${toolName}' not found`);
        return;
      }

      // Normalize parameters based on tool
      let normalizedParams = params;

      // Calculator: Handle "operands" array format by converting to a, b
      if (toolName === 'calculator' && params.operands && Array.isArray(params.operands)) {
        const [a, b] = params.operands as number[];
        normalizedParams = {
          operation: params.operation,
          a,
          b,
        };
      }

      console.log(`\n🔧 Using tool: ${toolName}`);
      console.log(`   Parameters: ${JSON.stringify(normalizedParams)}`);

      const result = await this.matimo.execute(toolName, normalizedParams);

      // Format result
      if (typeof result === 'object' && result !== null) {
        const resultData = result as any;
        if (resultData.stdout) {
          try {
            const parsed = JSON.parse(resultData.stdout);
            console.log(`\n✅ Result:`, parsed);
          } catch {
            console.log(`\n✅ Result:`, resultData.stdout);
          }
        } else if (resultData.data) {
          // HTTP response
          console.log(
            `\n✅ Result (HTTP ${resultData.statusCode}):`,
            typeof resultData.data === 'string'
              ? resultData.data.substring(0, 200)
              : JSON.stringify(resultData.data).substring(0, 200)
          );
        } else {
          console.log(`\n✅ Result:`, result);
        }
      }
    } catch (error) {
      console.error(`\n❌ Tool Error: ${error instanceof Error ? error.message : String(error)}`);
    }
  }
}

/**
 * Run factory pattern agent with prompts
 */
async function runFactoryPatternAgent() {
  console.log('\n╔════════════════════════════════════════════════════════╗');
  console.log('║   Matimo Factory Pattern - True AI Agent               ║');
  console.log('║   (AI decides which tool to use based on prompt)       ║');
  console.log('╚════════════════════════════════════════════════════════╝\n');

  try {
    const apiKey = process.env.OPENAI_API_KEY;
    if (!apiKey) {
      console.error('❌ Error: OPENAI_API_KEY not set in .env');
      process.exit(1);
    }

    // Initialize Matimo
    console.log('🚀 Initializing Matimo...');
    const matimo = await MatimoInstance.init({ autoDiscover: true });

    const matimoTools = matimo.listTools();
    console.log(`📦 Loaded ${matimoTools.length} tools:\n`);
    matimoTools.forEach((t) => {
      console.log(`  • ${t.name}`);
      console.log(`    ${t.description}\n`);
    });

    // Initialize OpenAI LLM
    console.log('🤖 Initializing OpenAI LLM (gpt-3.5-turbo)...\n');
    const llm = new ChatOpenAI({
      modelName: 'gpt-4o-mini',
      temperature: 0,
      openAIApiKey: apiKey,
    });

    // Create agent
    const agent = new FactoryPatternAgent(matimo, llm);

    // Test prompts - each should trigger a different tool
    const prompts = [
      '🧮 What is 42 plus 8?',
      '🔊 Echo the message: "Factory pattern works perfectly!"',
      '🌐 Fetch the GitHub user profile for octocat using HTTP GET',
    ];

    console.log('🧪 Testing AI Agent with 3 Different Prompts');
    console.log('═'.repeat(60));

    for (const prompt of prompts) {
      await agent.process(prompt);
      console.log('\n' + '─'.repeat(60));
    }

    console.log('\n✅ Factory pattern AI agent test complete!\n');
  } catch (error) {
    console.error('❌ Agent failed:', error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

// Run the agent
runFactoryPatternAgent().catch(console.error);

File: python/examples/native/agents/factory_pattern_agent.py
"""
Matimo Factory Pattern — AI Agent with Tool Decision Making.
Mirrors: typescript/examples/tools/agents/factory-pattern-agent.ts

The agent receives a prompt, uses OpenAI to decide which tool to use,
then executes that tool via Matimo.

Run:
    cd python
    OPENAI_API_KEY=sk-... uv run python examples/native/agents/factory_pattern_agent.py
"""
from __future__ import annotations

import asyncio
import json
import os
import re
import sys
from typing import Any

from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage, SystemMessage

from matimo import Matimo


class FactoryPatternAgent:
    """
    Factory Pattern Agent — uses OpenAI to decide which tool to call,
    then executes it via matimo.execute().
    """

    def __init__(self, matimo: Matimo, llm: ChatOpenAI) -> None:
        self._matimo = matimo
        self._llm = llm

    def _get_tool_schemas(self) -> list[dict]:
        """Build OpenAI function schemas from Matimo tool definitions."""
        return [
            {
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description,
                    "parameters": {
                        "type": "object",
                        "properties": {
                            param_name: {
                                "type": param.type.value if hasattr(param.type, "value") else param.type,
                                "description": param.description or "",
                                **({"enum": param.enum} if param.enum else {}),
                            }
                            for param_name, param in (t.parameters or {}).items()
                        },
                        "required": [
                            name for name, p in (t.parameters or {}).items() if p.required
                        ],
                    },
                },
            }
            for t in self._matimo.list_tools()
        ]

    async def process(self, prompt: str) -> None:
        print(f'\n❓ Prompt: "{prompt}"')

        tool_schemas = self._get_tool_schemas()
        tool_summary = ", ".join(
            f"{s['function']['name']}: {s['function']['description']}"
            for s in tool_schemas
        )

        messages = [
            SystemMessage(
                content=(
                    "You are an AI assistant with access to tools. "
                    "Based on the user's request, decide which tool to use and extract the required parameters. "
                    'Respond with JSON: {"tool": "<tool_name>", "parameters": {...}}'
                )
            ),
            HumanMessage(
                content=(
                    f"User request: {prompt}\n\n"
                    f"Available tools: {tool_summary}\n\n"
                    'Respond with JSON: {"tool": "<tool_name>", "parameters": {...}}'
                )
            ),
        ]

        try:
            response = await self._llm.ainvoke(messages)
            content = response.content

            tool_name: str | None = None
            tool_params: dict | None = None

            if isinstance(content, str):
                try:
                    parsed = json.loads(content)
                    tool_name = parsed.get("tool") or parsed.get("function", {}).get("name")
                    tool_params = parsed.get("parameters") or parsed.get("function", {}).get("parameters") or parsed
                except json.JSONDecodeError:
                    match = re.search(r'\{[^{}]*"tool"[^{}]*\}', content)
                    if match:
                        try:
                            parsed = json.loads(match.group())
                            tool_name = parsed.get("tool")
                            tool_params = parsed.get("parameters")
                        except json.JSONDecodeError:
                            pass

            if tool_name and tool_params is not None:
                await self._execute_tool(tool_name, tool_params)
            else:
                print("\n⚠️  No tool call detected in response")
                print(f"Response: {str(content)[:200]}")

        except Exception as exc:
            print(f"\n❌ Error: {exc}")

    async def _execute_tool(self, tool_name: str, params: dict[str, Any]) -> None:
        tool = self._matimo.get_tool(tool_name)
        if not tool:
            print(f"\n❌ Tool '{tool_name}' not found")
            return

        # Calculator: normalise operands array → a, b
        if tool_name == "calculator" and isinstance(params.get("operands"), list):
            a, b = params["operands"][0], params["operands"][1]
            params = {"operation": params.get("operation"), "a": a, "b": b}

        print(f"\n🔧 Using tool: {tool_name}")
        print(f"   Parameters: {json.dumps(params)}")

        try:
            result = await self._matimo.execute(tool_name, params)

            if isinstance(result, dict):
                if "stdout" in result:
                    try:
                        print("\n✅ Result:", json.loads(result["stdout"]))
                    except (json.JSONDecodeError, TypeError):
                        print("\n✅ Result:", result["stdout"])
                elif "data" in result:
                    data_str = json.dumps(result["data"])[:200]
                    print(f"\n✅ Result (HTTP {result.get('status_code', '?')}): {data_str}")
                else:
                    print("\n✅ Result:", result)
            else:
                print("\n✅ Result:", result)

        except Exception as exc:
            print(f"\n❌ Tool Error: {exc}")


async def main() -> None:
    print("\n╔════════════════════════════════════════════════════════╗")
    print("║   Matimo Factory Pattern - True AI Agent               ║")
    print("║   (AI decides which tool to use based on prompt)       ║")
    print("╚════════════════════════════════════════════════════════╝\n")

    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key:
        print("❌ Error: OPENAI_API_KEY not set")
        sys.exit(1)

    print("🚀 Initializing Matimo...")
    matimo = await Matimo.init(auto_discover=True)

    tools = matimo.list_tools()
    print(f"📦 Loaded {len(tools)} tools:\n")
    for t in tools:
        print(f"  • {t.name}")
        print(f"    {t.description}\n")

    print("🤖 Initializing OpenAI LLM (gpt-4o-mini)...\n")
    llm = ChatOpenAI(model="gpt-4o-mini", temperature=0, api_key=api_key)

    agent = FactoryPatternAgent(matimo, llm)

    prompts = [
        "🧮 What is 42 plus 8?",
        '🔊 Echo the message: "Factory pattern works perfectly!"',
        "🌐 Fetch the GitHub user profile for octocat using HTTP GET",
    ]

    print("🧪 Testing AI Agent with 3 Different Prompts")
    print("═" * 60)

    for prompt in prompts:
        await agent.process(prompt)
        print("\n" + "─" * 60)

    print("\n✅ Factory pattern AI agent test complete!\n")


if __name__ == "__main__":
    asyncio.run(main())
